### PR TITLE
make log cached immutable after creation

### LINKED ISSUE #914: remove mutable static state from the whole codebase
mutable static state is a source of memory leaks and concurrency bugs on dataflow/spark. we need to get rid of all of it. In particular mutable static collections but mutated static arrays are also bad (can't make java arrays unmodifiable unfortunately)

### PR DESCRIPTION
**Pull request by akiezun**
fixes #914 

@lbergelson please have a look

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 1]
```diff
@@ -509,41 +503,46 @@ public static boolean goodProbability(final double result) {
...
+        return logFactorialCache.get(x);
     }
 
     /**
-     * Wrapper class so that the logFactorial array is only calculated if it's used
+     * The size of the precomputed cache of logs.
+     * The caches are immutable after creation and so it's no big deal that they are static.
      */
-    private static class LogFactorialCache {
+    private static final int PRECOMPUTED_LOGS = 10_000;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: I'm a bit worried that we might be losing the benefit of the caching.  Do we know that most of the expense is in the 0-10,000 range?  Would we do better by just baking a log table into the jar?

**Comment by lbergelson** [comment]
I'm worried that we're losing the point of caching these.  Do we have any idea how important this is to performance/ how many of the calls fall in the 0-9,999 range?

Would it be better to consider replacing with a threadsafe cache that actually does caching rather than a selection of precomputed values?  

We could potentially use one of the [guava caches](https://code.google.com/p/guava-libraries/wiki/CachesExplained) if we actually need large values cached.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 2]
```diff
@@ -509,41 +503,46 @@ public static boolean goodProbability(final double result) {
...
+    private static final LogCache logCache = new LogCache(PRECOMPUTED_LOGS);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: I would just inline this to avoid having another static field

**Comment by lbergelson** [comment]
The code looks good, one trivial comment on it.  I'm worried that it defeats the purpose of the cache though, see my comment above.

**Comment by lbergelson** [comment]
@akiezun Back to you.

**Comment by akiezun** [comment]
if this is so, i think we will need to rediscover it in a profiler. It will be super obvious. So i dont worry about it. It will be heavily used in the genotyper - we'll run the genotyper and see what the cache miss rate is and what values it wants to compute logs or factorials of

**Comment by lbergelson** [comment]
Merge when ready then.

**Pull request merged**